nm/doc/nm_mgr.1
.\" Automatically generated by Pod::Man 4.11 (Pod::Simple 3.35)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "NM_MGR 1"
.TH NM_MGR 1 "2022-11-03" "perl v5.30.0" "NM executables"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
nm_mgr \- Network Management server implementing the Asynchronous Management Protocol (AMP)
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBnm_mgr\fR [\fIoptions\fR] \fImanager eid\fR
.PP
The following options may be specified to customize behavior.  Use \*(L"nm_mgr \-\-help\*(R" for full usage information:
.IP "\-A" 4
.IX Item "-A"
Startup directly in the alternative Automator \s-1UI\s0 mode. This mode is
designed to provide a consistent line-based interface suitable for
automated scripting.  Type ? when active for usage details.
.IP "\-l" 4
.IX Item "-l"
If specified, enable file-based logging of Manager Activity on startup.  This can be toggled at any time from the main menu of the \s-1UI.\s0
.Sp
If logging is not enabled, the following arguments have no affect until enabled in \s-1UI.\s0
.IP "\-d" 4
.IX Item "-d"
Log each agent to a different directory.
.IP "\-L #" 4
.IX Item "-L #"
Specify maximum number of entries (reports+tables) per file before rotating.
.IP "\-D \s-1DIR\s0" 4
.IX Item "-D DIR"
\&\s-1NM\s0 logs will be placed in this directory.
.IP "\-r" 4
.IX Item "-r"
Log all received reports to file in text format (as shown in \s-1UI\s0).
.IP "\-t" 4
.IX Item "-t"
Log all received tables to file in text format (as shown in \s-1UI\s0).
.IP "\-T" 4
.IX Item "-T"
Log all transmitted message as ASCII-encoded \s-1CBOR HEX\s0 strings.
.IP "\-R" 4
.IX Item "-R"
Log all received messages as ASCII-encoded \s-1CBOR HEX\s0 strings.
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
Starts the \fBnm_mgr\fR application listening on \fImgr eid\fR for messages from \fBnm_agent\fR clients.  Specify \*(L"\-\-help\*(R" for full usage information.
.PP
An agent will automatically attempt to register with it's configured manager on startup.  Agents may also be added manually through the managers \s-1UI.\s0
.PP
The manager provides a text based \s-1UI\s0 as its primary interface.  The \s-1UI\s0 provides capabilities to list, register, or delete agents. It can view received reports and tables, and be used to send commands (ARIs) to registered agents.
.PP
An experimental \s-1REST API\s0 is available if built with the configuration option \*(L"\-\-enable\-rest\*(R".  The default configuration will be accessible at http://localhost:8089/nm/api.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
Asynchronous Management Protocol <https://datatracker.ietf.org/doc/draft-birrane-dtn-amp/>
.PP
\&\fBnm_agent\fR\|(1)


release-notes.txt
= Release Notes for ION 4.1.1 =

November 19, 2021

%%%%%%%%%%%
= GENERAL =
%%%%%%%%%%%

The ION (interplanetary overlay network) software is a suite of communication
protocol implementations designed to support mission operation communications
across an end-to-end interplanetary network, which might include on-board
(flight) subnets, in-situ planetary or lunar networks, proximity links, deep
space links, and terrestrial internets.  Included in the ION software
distribution are the following packages:

- ici (interplanetary communication infrastructure), a set of libraries
  that provide flight-software-compatible support for functions on which
  the other packages rely, such as dynamic memory management, non-volatile
  storage management, and inter-task communication via shared memory.
  The ici libraries are designed to make the porting of IPN software to
  multiple operating systems - Linux, VxWorks, Solaris, etc. - as easy as
  possible. Ici now includes zco (zero-copy object), a library that 
  minimizes the copying of application data as it is encapsulated in 
  multiple layers of protocol structure while traversing the protocol
  stack.

- bp (bundle protocol), an implementation of the Delay-Tolerant
  Networking (DTN) architecture's Bundle Protocol.

- ltp (licklider transmission protocol), a DTN convergence layer protocol
  for reliable transmission over links characterized by long or highly
  variable delay.

- dgr (datagram retransmission), an alternative implementation of ltp
  that is designed for use over the Internet protocol stack.  dgr
  implements congestion control and is designed for relatively high
  performance.

- bssp (bundle streaming service protocol), a DTN convergence layer protocol
  for reliable transmission of streaming data.

- ams - an implementation of the CCSDS Asynchronous Message Service.

- cfdp - a class-1 (Unacknowledged) implementation of the CCSDS File
  Delivery Protocol.

- bss - a Bundle Streaming Service (BSS) for disruption-tolerant reliable
  data streaming.  BSS supports real-time streaming applications by
  passing the bundle payloads to the associated application for immediate
  display of the most recent data while storing all bundle payloads
  received into a database for user-directed playback.

- dtpc (delay-tolerant payload conditioning), an application framework
  providing TCP-like support for end-to-end retransmission and data delivery
  in transmission order without duplication or omission.

Features included:

- cgr - Contract graph routing: a method of dynamic routing designed for
  space based applications of ION, but still usable for terrestrial
  applications. It computes routes using scheduled communication and deals
  with time-varying network topology.

- brs - Bundle relay service: provides interconnectivity between networks
  that do not allow servers (those behind NAT for example). For more
  information, check man brsscla and man brsccla.

%%%%%%%%%%%%%%%%%%%%%%%%%%
= NOTES ON RELEASE 4.1.1 =
%%%%%%%%%%%%%%%%%%%%%%%%%%

Beginning with the ION 4.1.1 release, we will no longer be distributing 
two separate .tar.gz files (one that builds with BP version 6 by default and 
one that builds with BP version 7 by default). Instead, we will distribute a
single ION .tar.gz file that builds with BP version 7 by default. If you wish 
to use ION with BP version 6, you can use use the --enable-bpv6 configure 
option. If you are building ION from source on Windows for BP version 6, you
will need to run the command "make bpv6".

After ION version 4, all BP version 6 related changes will be bug fixes
unless explicitly identified as otherwise.

Branches included in this release:

misc-4.1.1-fixes
  * Improve the killm script 
  * Update pilot delay and add rate control for bpdriver
  * Change SourceCustodyRequired to NoCustodyRequested for AMS
  * Add a version of the req-003-multicast test for BP version 7
  * Fix a bug in librams.c that caused AMS petition assert/cancel
    flip-flop
  * Various updates for Makefiles (autoconf and manual)

apl-4.1.1-security-patches
  * Added BPSec Python test suite
  * Added accomodation of NULL terminator in security policy parameters
  * Added checks for NULL pointer references when a BCB targets an 
    extension block
  * Added TEST_POINT macros to BIB and BCB code to log test events

last-pre-git-tweaks
  * Add a warning about potential forwarding loops
  * Fix a typo in biberc documentation

autoconf-updates-MacOS_M1-WSL
  * Fix line ending encoding error in install_macos_sysctl.sh script
  * Add support for building ION on MacOS using the new M1 processor
    * At this time, ION works correctly on the M! processor with MacOS
      version 11.6 (Big Sur) but does not run properly on version
      12.0.1 (Monterey)
  * Add support to configure ION to build on WSL2 but not WSL1
  * ION has also now been tested on RedHat Linux


%%%%%%%%%%%%%%%%%%%%%%%%%%
= NOTES ON RELEASE 4.1.0 =
%%%%%%%%%%%%%%%%%%%%%%%%%%

ION 4.1.0 is the first release in which the BPv7 implementation is
conformant to the BPv7 protocol specification that will shortly be
published by IETF as a standards-track RFC.

The accompanying BPsec implementation is similarly conformant to
the BPSec specification that will also soon be a standard-track RFC.
NOTE, though, that the TCP convergence-layer protocol implementation
in ION 4.1.0 still conforms to the original TCPCL specification (RFC
7242), NOT to the upcoming TCPCL standards-track RFC; that upgrade
will be supported in a future release.

Another key change in this release is that Aggregate Custody Signaling
is no longer enabled by default.  (Due to some new features in ACS
configuration, when ACS is enabled the bpv6 bpclock daemon will no
longer run successfully unless the ACS database is initialized.
Since not all ION bpv6 users currently initialize ACS in node startup,
disabling ACS by default seems to be the best way forward.)  If you

nm/mgr/nm_mgr.c
/******************************************************************************
 **                           COPYRIGHT NOTICE
 **      (c) 2011 The Johns Hopkins University Applied Physics Laboratory
 **                         All rights reserved.
 ******************************************************************************/

/*****************************************************************************
 ** \file nm_mgr.c
 **
 ** File Name: nm_mgr.c
 **
 ** Subsystem:
 **          Network Manager Application
 **
 ** Description: This file implements the DTNMP Manager user interface
 **
 ** Notes:
 **
 ** Assumptions:
 **
 ** Modification History:
 **  MM/DD/YY  AUTHOR          DESCRIPTION
 **  --------  ------------    ---------------------------------------------
 **  09/01/11  V. Ramachandran Initial Implementation (JHU/APL)
 **  08/19/13  E. Birrane      Documentation clean up and code review comments. (JHU/APL)
 **  08/21/16  E. Birrane      Update to AMP v02 (Secure DTN - NASA: NNX14CS58P)
 **  10/06/18  E. Birrane     Update to AMP v0.5 (JHU/APL)
 *****************************************************************************/

// Application headers.
#include <getopt.h>
#include "nm_mgr.h"
#include "nm_mgr_ui.h"
#include "metadata.h"

#ifdef USE_CIVETWEB
#include "nm_rest.h"
#endif

#include "agents.h"

#include "../shared/primitives/rules.h"



mgr_db_t gMgrDB;
iif_t ion_ptr;
int  gRunning;

char* mgr_parse_args(int argc, char* argv[]);
void mgr_print_usage(void);

/******************************************************************************
 *
 * \par Function Name: main
 *
 * \par Main agent processing function.
 *
 * \param[in]  argc    # command line arguments.
 * \param[in]  argv[]  Command-line arguments.
 *
 * \par Notes:
 *
 * Modification History:
 *  MM/DD/YY  AUTHOR         DESCRIPTION
 *  --------  ------------   ---------------------------------------------
 **  09/01/11  V. Ramachandran Initial Implementation
 **  08/20/13  E. Birrane      Code Review Updates
 **  10/06/18  E. Birrane      Update to AMP v0.5 (JHU/APL)
 *****************************************************************************/

int main(int argc, char *argv[])
{
    pthread_t rx_thr;
    pthread_t ui_thr;

#ifdef HAVE_MYSQL
    pthread_t db_thr;
    char db_thr_name[] = "db_thread";
#endif

    char rx_thr_name[]     = "rx_thread";
    char ui_thr_name[]     = "ui_thread";
    char daemon_thr_name[] = "run_daemon";
    char *mgr_eid;

    errno = 0;

    /* Initialize the non-volatile database.
     *   Note: Initializing the structure here allows some attributes to be pre-defined by
     *   command line parsing if not re-initialized later.
     */
    memset((char*) &(gMgrDB), 0, sizeof(gMgrDB));

    if (argc > 2)
    {
        // Assume argv[1] is required manager_eid
        mgr_eid = mgr_parse_args(argc, argv);
        if (mgr_eid == NULL)
        {
            mgr_print_usage();
            return 1;
        }

    }
    else if(argc != 2)
    {
        fprintf(stderr,"Invalid number of arguments for nm_mgr\n");
        mgr_print_usage();
        return 1;
    }
    else
    {
        mgr_eid = argv[1];
    }

    /* Indicate that the threads should run once started. */
    gRunning = 1;

    /* Initialize the AMP Manager. */
    if(mgr_init(mgr_eid) != AMP_OK)
    {
    	AMP_DEBUG_ERR("main","Can't init Manager.", NULL);
    	exit(EXIT_FAILURE);
    }

    AMP_DEBUG_INFO("main","Manager EID: %s", argv[1]);


    /* Spawn threads for receiving msgs, user interface, and db connection. */
    if(pthread_begin(&rx_thr, NULL, (void *)mgr_rx_thread, (void *)&gRunning, "nm_mgr_rx"))
    {
        AMP_DEBUG_ERR("main","Can't create pthread %s, errnor = %s",
        		        rx_thr_name, strerror(errno));
        exit(EXIT_FAILURE);
    }


    if(pthread_begin(&ui_thr, NULL, (void *)ui_thread, (void *)&gRunning, "nm_mgr_ui"))
    {
        AMP_DEBUG_ERR("main","Can't create pthread %s, errnor = %s",
        		        ui_thr_name, strerror(errno));
        exit(EXIT_FAILURE);
    }

#ifdef HAVE_MYSQL

    if(pthread_begin(&db_thr, NULL, (void *)db_mgt_daemon, (void *)&gRunning ,"nm_mgr_db"))
    {
    	AMP_DEBUG_ERR("main","Can't create pthread %s, errnor = %s",
    			db_thr_name, strerror(errno));
    	exit(EXIT_FAILURE);
    }
#endif

#ifdef USE_CIVETWEB
    nm_rest_start();
#endif

    if (pthread_join(rx_thr, NULL))
    {
        AMP_DEBUG_ERR("main","Can't join pthread %s. Errnor = %s",
        		        rx_thr_name, strerror(errno));
        exit(EXIT_FAILURE);
    }
    if (pthread_join(ui_thr, NULL))
    {
        AMP_DEBUG_ERR("main","Can't join pthread %s. Errnor = %s",
        		         ui_thr_name, strerror(errno));
        exit(EXIT_FAILURE);
    }

#ifdef USE_CIVETWEB
    nm_rest_stop();
#endif
    
#ifdef HAVE_MYSQL
    if (pthread_join(db_thr, NULL))
    {
    	AMP_DEBUG_ERR("main","Can't join pthread %s. Errnor = %s",
    			        db_thr_name, strerror(errno));
    	exit(EXIT_FAILURE);
    }
#endif

    AMP_DEBUG_ALWAYS("main","Shutting down manager.", NULL);
    mgr_cleanup();

    AMP_DEBUG_INFO("main","Exiting Manager after cleanup.", NULL);
    exit(0);
}




/******************************************************************************
 *
 * \par Function Name: mgr_cleanup
 *
 * \par Cleans resources before exiting the manager.
 *
 * \par Notes:
 *
 * Modification History:
 *  MM/DD/YY  AUTHOR         DESCRIPTION
 *  --------  ------------   ---------------------------------------------
 **  09/01/11  V. Ramachandran Initial Implementation
 **  08/20/13  E. Birrane      Code Review Updates
 **  10/06/18  E. Birrane      Updated to AMP v0.5 (JHU/APL)
 *****************************************************************************/

int mgr_cleanup()
{

#ifdef HAVE_MYSQL
	db_mgt_close();
#endif

	vec_release(&(gMgrDB.agents), 0);
	rhht_release(&(gMgrDB.metadata), 0);

	db_destroy();

	utils_mem_teardown();

	return AMP_OK;
}



/******************************************************************************
 *
 * \par Function Name: mgr_init
 *
 * \par Initialize the manager...
 *
 * \par Notes:
 *
 * Modification History:
 *  MM/DD/YY  AUTHOR         DESCRIPTION
 *  --------  ------------   ---------------------------------------------
 **  09/01/11  V. Ramachandran Initial Implementation
 **  08/20/13  E. Birrane      Code Review Updates
 **  10/06/18  E. Birrane      Update to AMP v0.5 (JHU/APL)
 *****************************************************************************/
int mgr_init(char *arg_eid)
{
	int success;

	AMP_DEBUG_ENTRY("mgr_init","("ADDR_FIELDSPEC")",(uaddr) arg_eid);

    /* Step 2: Make sure that ION is running and we can attach. */
	if (ionAttach() < 0)
	{
		AMP_DEBUG_ERR("mgr_init", "Manager can't attach to ION.", NULL);
		return -1;
	}


	/* Step 1: Initialize MGR-specific data.*/
	gMgrDB.agents = vec_create(AGENT_DEF_NUM_AGTS, agent_cb_del,agent_cb_comp, NULL, 0, &success);
	if(success != VEC_OK)
	{
		AMP_DEBUG_ERR("mgr_init", "Can't make agents vec.", NULL);
		return AMP_FAIL;
	}

	gMgrDB.metadata = rhht_create(NM_MGR_MAX_META, ari_cb_comp_no_parm_fn, ari_cb_hash, meta_cb_del, &success);
	if(success != RH_OK)
	{
		AMP_DEBUG_ERR("mgr_init", "Can't make parmspec ht.", NULL);
		return AMP_FAIL;
	}


	gMgrDB.tot_rpts = 0;
    gMgrDB.tot_tbls = 0;
    istrcpy((char *) gMgrDB.mgr_eid.name, arg_eid, AMP_MAX_EID_LEN);


	/* Step 2:  Attach to ION. */
    if(iif_register_node(&ion_ptr, gMgrDB.mgr_eid) == 0)
    {
        AMP_DEBUG_ERR("mgr_init","Unable to register BP Node. Exiting.", NULL);
        return AMP_FAIL;
    }

    if (iif_is_registered(&ion_ptr))
    {
        AMP_DEBUG_INFO("mgr_init", "Mgr registered with ION, EID: %s",
        		         iif_get_local_eid(&ion_ptr).name);
    }
    else
    {
        AMP_DEBUG_ERR("mgr_init","Failed to register mgr with ION, EID %s",
        				 iif_get_local_eid(&ion_ptr).name);
        AMP_DEBUG_EXIT("mgr_init","->-1.",NULL);
        return AMP_FAIL;
    }


    if((utils_mem_int()       != AMP_OK) ||
       (db_init("nmmgr_db", &adm_init) != AMP_OK))
    {
    	db_destroy();
    	AMP_DEBUG_ERR("mgr_init","Unable to initialize DB.", NULL);
    	return AMP_FAIL;
    }

#ifdef HAVE_MYSQL
	db_mgr_sql_init();
	success = db_mgt_init(gMgrDB.sql_info, 0, 1);
#endif

    success = AMP_OK;

    return success;
}

/**
 * Parse optional command line arguments
 */
char* mgr_parse_args(int argc, char* argv[])
{
    int i;
    int c;
    int option_index = 0;
    static struct option long_options[] =
        {
            {"log", no_argument, 0,'l'},
            {"log-to-dirs", no_argument, 0,'d'},
            {"log-rx-rpt", no_argument, 0,'r'},
            {"log-rx-tbl", no_argument, 0,'t'},
            {"log-tx-cbor", no_argument, 0,'T'},
            {"log-rx-cbor", no_argument, 0,'R'},
            {"log-tx-cbor", no_argument, 0,'j'},
            {"log-rx-cbor", no_argument, 0,'J'},
            
            {"sql-user", required_argument, 0,'u'},
            {"sql-pass", required_argument, 0,'p'},
            {"sql-db", required_argument,0, 'S'},
            {"sql-host", required_argument,0, 's'},
            
            {"log-dir", required_argument, 0,'D'},
            {"log-limit", required_argument, 0,'L'},
            {"automator", required_argument, 0,'a'},
            {"help", required_argument, 0,'h'},
        };
    while ((c = getopt_long(argc, argv, "ldL:D:rtTRaAjJs:u:p:S:", long_options, &option_index)) != -1)
    {
        switch(c)
        {
        case 'l':
            agent_log_cfg.enabled = 1;
            break;
        case 'd':
            agent_log_cfg.agent_dirs = 1;
            break;
        case 'r':
            agent_log_cfg.rx_rpt = 1;
            break;
        case 't':
            agent_log_cfg.rx_tbl = 1;
            break;
        case 'T':
            agent_log_cfg.tx_cbor = 1;
            break;
        case 'R':
            agent_log_cfg.rx_cbor = 1;
            break;
#ifdef USE_JSON
        case 'j':
            agent_log_cfg.rx_json_rpt = 1;
            break;
        case 'J':
            agent_log_cfg.rx_json_tbl = 1;
            break;
#endif
#ifdef HAVE_MYSQL
        case 's': // MySQL Server
            strncpy(gMgrDB.sql_info.server, optarg, UI_SQL_SERVERLEN-1);
            break;
        case 'u': // MySQL Username
            strncpy(gMgrDB.sql_info.username, optarg, UI_SQL_ACCTLEN-1);
            break;
        case 'p': // MySQL Password
            strncpy(gMgrDB.sql_info.password, optarg, UI_SQL_ACCTLEN-1);
            break;
        case 'S': // MySQL Database Name
            strncpy(gMgrDB.sql_info.database, optarg, UI_SQL_DBLEN-1);
            break;

#endif
        case 'D':
            strncpy(agent_log_cfg.dir, optarg, sizeof(agent_log_cfg.dir)-1);
            break;
        case 'L':
            agent_log_cfg.limit = atoi(optarg);
            break;
        case 'a':
        case 'A':
            mgr_ui_mode = MGR_UI_AUTOMATOR;
            break;
        case 'h':
            return NULL;
        default:
            fprintf(stderr, "Error parsing arguments\n");
            return NULL;
        }
    }

    // Check for any remaining unrecognized arguments
    if ((argc-optind) != 1)
    {
        fprintf(stderr,"%d unrecognized arguments:\n", (argc-optind));
        for(i = optind; i < argc; i++)
        {
            printf("\t%s\n", argv[i]);
        }
        return NULL;
    }
    else
    {
        return argv[optind];
    }
}

void mgr_print_usage(void)
{

    printf("AMP Protocol Version %d - %s, built on %s %s\n",
           AMP_VERSION,
           AMP_PROTOCOL_URL,
           __DATE__, __TIME__);

    printf("Usage: nm_mgr [options] <manager eid>\n");
    printf("Supported Options:\n");
    printf("-A       Startup directly in the alternative Automator UI mode\n");
    printf("-l       If specified, enable file-based logging of Manager Activity.\n");
    printf("           If logging is not enabled, the following have no affect until enabled in UI\n");
    printf("-d       Log each agent to a different directory\n");
    printf("-L #      Specify maximum number of entries (reports+tables) per file before rotating\n");
    printf("-D DIR   NM logs will be placed in this directory\n");
    printf("-r       Log all received reports to file in text format (as shown in UI)\n");
    printf("-t       Log all received tables to file in text format (as shown in UI)\n");
    printf("-T       Log all transmitted message as ASCII-encoded CBOR HEX strings\n");
    printf("-R       Log all received messages as ASCII-encoded CBOR HEX strings\n");
#ifdef HAVE_MYSQL
    printf("--sql-user MySQL Username\n");
    printf("--sql-pass MySQL Password\n");
    printf("--sql-db MySQL Datbase Name\n");
    printf("--sql-host MySQL Host\n");
#endif
}
